syntax = "proto3";

package prism;

option go_package = "github.com/sjanney/prism/proto";

service PrismService {
  rpc Index(IndexRequest) returns (stream IndexProgress) {}
  rpc Search(SearchRequest) returns (SearchResponse) {}
  rpc OpenResult(OpenRequest) returns (OpenResponse) {}
  rpc ConnectDatabase(ConnectDatabaseRequest) returns (ConnectDatabaseResponse) {}
  rpc GetStats(GetStatsRequest) returns (GetStatsResponse) {}
  rpc GetSystemInfo(GetSystemInfoRequest) returns (GetSystemInfoResponse) {}
  rpc ActivateLicense(ActivateLicenseRequest) returns (ActivateLicenseResponse) {}
  rpc PickFolder(PickFolderRequest) returns (PickFolderResponse) {}
  // Benchmark & Diagnostics
  rpc RunBenchmark(RunBenchmarkRequest) returns (stream BenchmarkProgress) {}
  rpc GetBenchmarkReport(GetBenchmarkReportRequest) returns (BenchmarkReport) {}
  
  // Cloud Credentials
  rpc SaveCloudCredentials(SaveCloudCredentialsRequest) returns (SaveCloudCredentialsResponse) {}
  rpc GetCloudCredentials(GetCloudCredentialsRequest) returns (GetCloudCredentialsResponse) {}
  rpc ValidateCloudCredentials(ValidateCloudCredentialsRequest) returns (ValidateCloudCredentialsResponse) {}
}


message GetSystemInfoRequest {}

message GetSystemInfoResponse {
  string device = 1;
  string siglip_model = 2;
  string yolo_model = 3;
  string backend_version = 4;
  int32 cpu_count = 5;
  string memory_usage = 6;
  bool is_pro = 7;
  bool developer_mode = 8;
  string license_email = 9;
  string license_expires = 10;
}


message ConnectDatabaseRequest {
  string db_path = 1;
}

message ConnectDatabaseResponse {
  bool success = 1;
  string message = 2;
}

message IndexRequest {
  string path = 1;
}

message IndexProgress {
  int64 current = 1;
  int64 total = 2;
  string status_message = 3;
  int64 skipped = 4;        // Count of skipped (duplicate) files
  int32 eta_seconds = 5;    // Estimated time remaining
}

message SearchRequest {
  string query_text = 1;
  float min_confidence = 2;       // Optional: minimum confidence filter
  repeated string object_filter = 3; // Optional: filter by detected objects
  int32 limit = 4;                // Optional: max results to return
}

message SearchResult {
  string path = 1;
  float confidence = 2;
  string reasoning = 3;
  // New fields for UI details
  string resolution = 4;
  string date_modified = 5;
  string file_size = 6;
  repeated string detected_objects = 7;
  string match_type = 8;          // "full_image" or "object_crop"
}

message SearchResponse {
  repeated SearchResult results = 1;
  int32 search_time_ms = 2;       // Search duration in milliseconds
  int32 total_count = 3;          // Total number of results
}

message OpenRequest {
  string file_path = 1;
}

message OpenResponse {
  bool success = 1;
  string message = 2;
}

message GetStatsRequest {}

message DatasetMetadata {
  int64 total_frames = 1;
  int64 total_embeddings = 2;
  string last_indexed = 3;
  string db_path = 4;
}

message GetStatsResponse {
  DatasetMetadata metadata = 1;
}

message ActivateLicenseRequest {
  string license_key = 1;
}

message ActivateLicenseResponse {
  bool success = 1;
  string message = 2;
}

message PickFolderRequest {
  string prompt = 1;
}

message PickFolderResponse {
  bool success = 1;
  string path = 2;
  string message = 3;
}

// ==================== Benchmark & Diagnostics ====================

message RunBenchmarkRequest {
  string sample_path = 1;  // Path to sample images (default: data/sample)
}

message BenchmarkProgress {
  string phase = 1;         // "indexing" | "search" | "complete"
  int32 current = 2;
  int32 total = 3;
  string message = 4;
}

message BenchmarkMetric {
  string name = 1;
  double value = 2;
  string unit = 3;
  string context = 4;       // e.g., "batch_size=8, device=mps"
}

message BenchmarkReport {
  string timestamp = 1;
  string prism_version = 2;
  string device = 3;
  string os = 4;
  repeated BenchmarkMetric indexing_metrics = 5;
  repeated BenchmarkMetric search_metrics = 6;
  repeated BenchmarkMetric system_metrics = 7;
}

message SaveCloudCredentialsRequest {
  string provider = 1; // "aws" or "azure"
  string aws_access_key = 2;
  string aws_secret_key = 3;
  string aws_region = 4;
  string azure_connection_string = 5;
}

message SaveCloudCredentialsResponse {
  bool success = 1;
  string message = 2;
}

message GetCloudCredentialsRequest {
  string provider = 1;
}

message GetCloudCredentialsResponse {
  bool configured = 1;
  string aws_access_key_masked = 2; // e.g., "****ABCD"
  string aws_region = 3;
  string azure_account_name = 4; // Extracted from conn string
}

message ValidateCloudCredentialsRequest {
  string provider = 1;
}

message ValidateCloudCredentialsResponse {
  bool success = 1;
  string message = 2;
}

message GetBenchmarkReportRequest {}

